### PR TITLE
bmips: update ethernet driver

### PR DESCRIPTION
**Pull request by spliew89**
Improve performance and add features to the bmips ethernet driver to bring it to
parity with bcm63xx.
- don't use NET_IP_ALIGN in copybreak which messes up alignment with DSA tag
- convert to napi_build_skb()
- use rx batch processing
- add BQL support

Tested on BCM6328 using iperf3.
The use of DSA in bmips actually lowered throughput to ~47Mbps from the max
100Mbps under bcm63xx. This series of changes uplift performance by ~20% to
~54Mbps (with flow offload ~74Mbps).

### CODE REVIEW

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c [hunk 1]
```diff
@@ -251,32 +254,35 @@ static inline void dmas_writel(struct bcm6368_enetsw *priv, u32 val,
...
 		desc_idx = priv->rx_dirty_desc;
 		desc = &priv->rx_desc_cpu[desc_idx];
 
-		if (!priv->rx_skb[desc_idx]) {
-			skb = netdev_alloc_skb(dev, priv->rx_skb_size);
-			if (!skb)
+		if (!priv->rx_buf[desc_idx]) {
+			unsigned char *buf;
+
+			if(likely(napi_mode))
```
Comment (Ansuel) [inline target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:272]: nitpick space after if.

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c [hunk 2]
```diff
@@ -251,32 +254,35 @@ static inline void dmas_writel(struct bcm6368_enetsw *priv, u32 val,
...
+				buf = napi_alloc_frag(priv->rx_frag_size);
+			else
+				buf = netdev_alloc_frag(priv->rx_frag_size);
```
Comment (Ansuel) [inline target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:275]: nitpick extra space after else condition ending

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c [hunk 3]
```diff
@@ -251,32 +254,35 @@ static inline void dmas_writel(struct bcm6368_enetsw *priv, u32 val,
...
+			if (unlikely(!buf))
 				break;
-			priv->rx_skb[desc_idx] = skb;
-			p = dma_map_single(&priv->pdev->dev, skb->data,
-					   priv->rx_skb_size,
-					   DMA_FROM_DEVICE);
-			desc->address = p;
+			priv->rx_buf[desc_idx] = buf;
+			desc->address = dma_map_single(&priv->pdev->dev,
+										   buf + NET_SKB_PAD,
```
Comment (Ansuel) [inline target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:280]: what happen here with the spacing?

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c [hunk 4]
```diff
@@ -365,43 +375,57 @@ static int bcm6368_enetsw_receive_queue(struct net_device *dev, int budget)
...
 		/* don't include FCS */
 		len -= 4;
 
 		if (len < priv->copybreak) {
-			struct sk_buff *nskb;
-
-			nskb = napi_alloc_skb(&priv->napi, len);
-			if (!nskb) {
+			unsigned int nfrag_size = SKB_DATA_ALIGN(NET_SKB_PAD + len +
+													 SKB_DATA_ALIGN(sizeof(struct skb_shared_info)));
```
Comment (Ansuel) [inline target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:386]: extra spacing?

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c [hunk 5]
```diff
@@ -365,43 +375,57 @@ static int bcm6368_enetsw_receive_queue(struct net_device *dev, int budget)
...
 						   len, DMA_FROM_DEVICE);
-			skb = nskb;
+			buf = nbuf;
+			frag_size = nfrag_size;
 		} else {
-			dma_unmap_single(&priv->pdev->dev, desc->address,
-					 priv->rx_skb_size, DMA_FROM_DEVICE);
-			priv->rx_skb[desc_idx] = NULL;
+			dma_unmap_single(kdev, desc->address,
+							 priv->rx_buf_size, DMA_FROM_DEVICE);
```
Comment (Ansuel) [inline target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:403]: wrong spacing?

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c [hunk 6]
```diff
@@ -960,9 +992,12 @@ static int bcm6368_enetsw_probe(struct platform_device *pdev)
 		dev_info(dev, "random mac %pM\n", ndev->dev_addr);
 	}
 
-	priv->rx_skb_size = ALIGN(ndev->mtu + ENETSW_MTU_OVERHEAD,
+	priv->rx_buf_size = ALIGN(ndev->mtu + ENETSW_MTU_OVERHEAD,
 				  priv->dma_maxburst * 4);
 
+	priv->rx_frag_size = SKB_DATA_ALIGN(NET_SKB_PAD + priv->rx_buf_size +
+										SKB_DATA_ALIGN(sizeof(struct skb_shared_info)));
```
Comment (Ansuel) [inline target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:999]: wrong spacing?

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c (hunk 3 above)
Comment (spliew89) [reply target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:280]: Sorry, vscode messes up the spacing sometimes.

The bcm63xx_net driver, which this driver is based on, doesn't check the dma addr.
I can add dma_mapping_error() checking if needed.

Comment (Ansuel) [reply target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:280]: yes pls add the dma check
Np about the space it was just to warn you on what to fix :D

Comment (Ansuel) [reply target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:280]: (if you want you can put the dma_mapping_error() in a separate commit)

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c (hunk 1 above)
Comment (spliew89) [reply target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:273]: Done

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c (hunk 2 above)
Comment (spliew89) [reply target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:276]: Done

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c (hunk 1 above)
Comment (Ansuel) [reply target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c:273]: Mhhh think you misunderstood :D
Was referring to
```
if(likely(napi_mode))
```
that should be changed to
```
if (likely(napi_mode))
```

not adding an extra space after
```
if (!priv->rx_buf[desc_idx]) {
```
(that was fine)

**Comment by Ansuel** [comment]
@spliew89 thanks for the changes but you should fix the related commits instead of creating an additional commit that fix the spacing.
(so use git rebase, edit the related commit with the wrong spacing and then force push)

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c (hunk 3 above)
Comment (spliew89) [reply target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c]: > (if you want you can put the dma_mapping_error() in a separate commit)

https://github.com/openwrt/openwrt/pull/10144/commits/5cf4fccdecc8f980e3bf3284781d2e013e4a0dba

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c (hunk 1 above)
Comment (spliew89) [reply target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c]: Done

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c (hunk 3 above)
Comment (spliew89) [reply target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c]: Done

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c (hunk 4 above)
Comment (spliew89) [reply target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c]: Done

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c (hunk 5 above)
Comment (spliew89) [reply target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c]: Done

File: target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c (hunk 6 above)
Comment (spliew89) [reply target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c]: Done

**Comment by Noltari** [comment]
Merged, thanks @spliew89!
https://github.com/Noltari/openwrt/compare/3817577c92618256565a55213e440cbcd1bd4e86...65ebe311304a32423a46bf4efe62f29cda06acab

**Pull request closed without merging**